{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c901d76b-0dc3-44f4-9497-24162b68f236",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.16 (BiocManager 1.30.19), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'BayesSpace'\n",
      "\n",
      "also installing the dependencies ‘miscTools’, ‘sandwich’, ‘plogr’, ‘statmod’, ‘bluster’, ‘metapod’, ‘Formula’, ‘maxLik’, ‘RSQLite’, ‘filelock’, ‘scran’, ‘coda’, ‘mclust’, ‘DirichletReg’, ‘xgboost’, ‘BiocFileCache’, ‘RcppDist’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'batchelor', 'BH', 'BiocParallel', 'boot', 'broom', 'bslib',\n",
      "  'class', 'cli', 'codetools', 'colorspace', 'curl', 'dbplyr', 'digest',\n",
      "  'dplyr', 'e1071', 'edgeR', 'evaluate', 'fansi', 'fontawesome', 'forcats',\n",
      "  'foreign', 'formatR', 'fs', 'future', 'gargle', 'GenomeInfoDb',\n",
      "  'GenomicRanges', 'gert', 'ggbeeswarm', 'ggrepel', 'glmnet', 'gower', 'hdf5r',\n",
      "  'highr', 'htmltools', 'htmlwidgets', 'httpuv', 'IRkernel', 'isoband',\n",
      "  'knitr', 'lava', 'leidenbase', 'limma', 'listenv', 'locfit', 'lubridate',\n",
      "  'MASS', 'MAST', 'Matrix', 'modelr', 'nlme', 'parallelly', 'pbapply',\n",
      "  'pbdZMQ', 'pkgbuild', 'pkgdown', 'pkgload', 'progressr', 'pryr', 'purrr',\n",
      "  'ragg', 'raster', 'Rcpp', 'RcppArmadillo', 'RcppTOML', 'RCurl', 'recipes',\n",
      "  'repr', 'reticulate', 'rgeos', 'rmarkdown', 'roxygen2', 'rsample', 's2',\n",
      "  'S4Vectors', 'sass', 'scuttle', 'sf', 'shiny', 'slider', 'sourcetools', 'sp',\n",
      "  'spatial', 'spatstat.explore', 'spatstat.geom', 'spatstat.random', 'spData',\n",
      "  'stringi', 'survival', 'terra', 'testthat', 'tidyr', 'timechange',\n",
      "  'timeDate', 'tinytex', 'units', 'utf8', 'vctrs', 'vroom', 'whisker', 'wk',\n",
      "  'xfun', 'yaml'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"BayesSpace\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9fca7a-8d90-41e4-9f4f-0c86eccb0fe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Author    : A. Alsema\n",
    "# Date      : Augustus 2021\n",
    "# Dataset   : Visium Spatial Transcriptomics for MS lesions\n",
    "# Purpose   : create SCE for spatial clustering with BayesSpace\n",
    "# Output    : \n",
    "# - sampleID_sce.rds, a SingleCellExperiment object class for each sample.\n",
    "# - plot of the different q-values that helps to set q\n",
    "# Required inputs: \n",
    "# - inpath: directory with spaceranger outputs. Expects that each sample has it's own subdirectory called \"sampleID-out\"\n",
    "#########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ecb2431f-e011-482a-985c-8d509e82aabb",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(BayesSpace)\n",
    "rm(list = ls())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "560bca07-ebd5-42d7-a417-04b35a39955c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dir.create(\"Routput/BS/qplots\", recursive = TRUE)\n",
    "dir.create(\"RData/BS/1.persample\", recursive = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2055b2b0-4b4f-415b-8bd4-cb2406d2a162",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST31\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3550\n",
      "[1] 33538  3550\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3550 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST32\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3067\n",
      "[1] 33538  3067\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3067 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST33\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  2961\n",
      "[1] 33538  2897\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 2897 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST34\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4284\n",
      "[1] 33538  4147\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4147 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST37\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4436\n",
      "[1] 33538  4436\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4436 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST38\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3797\n",
      "[1] 33538  3797\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3797 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST67\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4208\n",
      "[1] 33538  4208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4208 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST68\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3649\n",
      "[1] 33538  3648\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3648 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST69\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4789\n",
      "[1] 33538  4789\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4789 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST70\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3627\n",
      "[1] 33538  3627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3627 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST71\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3657\n",
      "[1] 33538  3657\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3657 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST72\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4895\n",
      "[1] 33538  4895\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4895 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST73\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  3751\n",
      "[1] 33538  3751\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 3751 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST74\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4702\n",
      "[1] 33538  4702\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4702 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ST79\"\n",
      "[1] \"dimensions of your sample\"\n",
      "[1] 33538  4729\n",
      "[1] 33538  4729\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Neighbors were identified for 0 out of 4729 spots.\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n",
      "Fitting model...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Vector of WM sample IDs to be processed\n",
    "sampleIDs <- c(\"ST31\", \"ST32\", \"ST33\", \"ST34\", \"ST37\", \"ST38\", \"ST67\", \"ST68\", \"ST69\", \"ST70\", \"ST71\", \"ST72\", \"ST73\", \"ST74\", \"ST79\")\n",
    "# identical procedure for GM\n",
    "\n",
    "# Loop over each sample ID\n",
    "for (i in 1:length(sampleIDs)) {\n",
    "    # Extract the current sample ID\n",
    "    sampleID <- sampleIDs[i]\n",
    "    print(sampleID)\n",
    "    \n",
    "    # Read in the Visium data for the current sample\n",
    "    sce <- readVisium(paste0(inpath, \"/spaceranger_output/\", sampleID, \"-out/outs/\"))\n",
    "    print(\"dimensions of your sample\")\n",
    "    print(dim(sce))\n",
    "    \n",
    "    # Remove spots with zero counts\n",
    "    sce <- sce[, colSums(counts(sce)) > 0]\n",
    "    \n",
    "    # Print dimensions of the sample data after removing zero-count spots\n",
    "    print(\"dimensions of your sample after filter\")\n",
    "    print(dim(sce))\n",
    "    \n",
    "    # Log-normalize the count matrix and perform PCA on the top 2000 highly variable genes (HVGs)\n",
    "    sce <- spatialPreprocess(sce, platform = \"Visium\", n.HVGs = 2000)\n",
    "    \n",
    "    # Evaluate clustering resolutions between 2 and 15\n",
    "    sce <- qTune(sce, qs = seq(2, 15), platform = \"ST\", d = 15)\n",
    "    \n",
    "    # Generate and save a q-value plot for the different resolutions\n",
    "    tiff(paste0(\"Routput/BS/qplots/resolution_qvalue_\", sampleID, \".tiff\"))\n",
    "    print(qPlot(sce, qs = seq(2, 15)))\n",
    "    dev.off()\n",
    "    \n",
    "    # Save the processed single-cell experiment object to an RDS file\n",
    "    saveRDS(sce, paste0(\"RData/BS/1.persample/\", sampleID, \"_sce.rds\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7f2af52-18c5-4041-9d20-e83e968fd8bc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.2.0 (2022-04-22)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/bcn/p283607/anaconda3/envs/R4.2/lib/libopenblasp-r0.3.21.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] BayesSpace_1.8.2            SingleCellExperiment_1.20.0\n",
       " [3] SummarizedExperiment_1.28.0 Biobase_2.58.0             \n",
       " [5] GenomicRanges_1.50.0        GenomeInfoDb_1.34.1        \n",
       " [7] IRanges_2.32.0              S4Vectors_0.36.0           \n",
       " [9] BiocGenerics_0.44.0         MatrixGenerics_1.10.0      \n",
       "[11] matrixStats_0.63.0          dplyr_1.0.10               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] bitops_1.0-7              bit64_4.0.5              \n",
       "  [3] filelock_1.0.2            httr_1.4.4               \n",
       "  [5] repr_1.1.4                tools_4.2.0              \n",
       "  [7] utf8_1.2.2                R6_2.5.1                 \n",
       "  [9] irlba_2.3.5.1             vipor_0.4.5              \n",
       " [11] DirichletReg_0.7-1        DBI_1.1.3                \n",
       " [13] colorspace_2.0-3          rhdf5filters_1.10.0      \n",
       " [15] tidyselect_1.2.0          gridExtra_2.3            \n",
       " [17] bit_4.0.5                 curl_4.3.3               \n",
       " [19] compiler_4.2.0            cli_3.4.1                \n",
       " [21] BiocNeighbors_1.16.0      DelayedArray_0.24.0      \n",
       " [23] sandwich_3.0-2            scales_1.2.1             \n",
       " [25] rappdirs_0.3.3            pbdZMQ_0.3-8             \n",
       " [27] digest_0.6.30             XVector_0.38.0           \n",
       " [29] base64enc_0.1-3           scater_1.26.1            \n",
       " [31] pkgconfig_2.0.3           htmltools_0.5.3          \n",
       " [33] sparseMatrixStats_1.10.0  limma_3.54.0             \n",
       " [35] dbplyr_2.2.1              fastmap_1.1.0            \n",
       " [37] rlang_1.0.6               RSQLite_2.2.20           \n",
       " [39] DelayedMatrixStats_1.20.0 generics_0.1.3           \n",
       " [41] zoo_1.8-11                jsonlite_1.8.4           \n",
       " [43] mclust_6.0.0              BiocParallel_1.32.0      \n",
       " [45] RCurl_1.98-1.9            magrittr_2.0.3           \n",
       " [47] BiocSingular_1.14.0       GenomeInfoDbData_1.2.9   \n",
       " [49] Formula_1.2-4             scuttle_1.8.0            \n",
       " [51] Matrix_1.5-1              Rcpp_1.0.9               \n",
       " [53] ggbeeswarm_0.6.0          IRkernel_1.3.1           \n",
       " [55] munsell_0.5.0             Rhdf5lib_1.20.0          \n",
       " [57] fansi_1.0.3               viridis_0.6.2            \n",
       " [59] lifecycle_1.0.3           edgeR_3.40.0             \n",
       " [61] zlibbioc_1.44.0           rhdf5_2.42.0             \n",
       " [63] BiocFileCache_2.6.0       grid_4.2.0               \n",
       " [65] blob_1.2.3                dqrng_0.3.0              \n",
       " [67] parallel_4.2.0            ggrepel_0.9.2            \n",
       " [69] crayon_1.5.2              lattice_0.20-45          \n",
       " [71] IRdisplay_1.1             beachmat_2.14.0          \n",
       " [73] locfit_1.5-9.6            metapod_1.6.0            \n",
       " [75] pillar_1.8.1              igraph_1.3.5             \n",
       " [77] uuid_1.1-0                xgboost_1.7.3.1          \n",
       " [79] codetools_0.2-18          ScaledMatrix_1.6.0       \n",
       " [81] glue_1.6.2                evaluate_0.18            \n",
       " [83] data.table_1.14.6         scran_1.26.2             \n",
       " [85] BiocManager_1.30.19       vctrs_0.5.1              \n",
       " [87] miscTools_0.6-26          gtable_0.3.1             \n",
       " [89] purrr_0.3.5               assertthat_0.2.1         \n",
       " [91] cachem_1.0.6              ggplot2_3.4.0            \n",
       " [93] rsvd_1.0.5                coda_0.19-4              \n",
       " [95] viridisLite_0.4.1         tibble_3.1.8             \n",
       " [97] beeswarm_0.4.0            memoise_2.0.1            \n",
       " [99] cluster_2.1.4             maxLik_1.5-2             \n",
       "[101] statmod_1.5.0             bluster_1.8.0            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R4.2",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
