{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8f70bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Author    : A. Alsema, J. Kotah\n",
    "# Date      : October 2023\n",
    "# Dataset   : Visium Spatial Transcriptomics for MS lesions, 14 WM samples\n",
    "# Purpose   : run first half of SpaceX analysis (until post-processing)\n",
    "# Output    : RDS SpaceX analyzed object per sample that needs post-processing\n",
    "# Input     : SpaceRanger output data files per sample, list of spatial coordinates, list of spatially variable genes per sample\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7efacfa",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: PQLseq\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2024-02-28 14:35:56 CET\"\n"
     ]
    }
   ],
   "source": [
    "rm(list = ls())\n",
    "library(SpaceX)\n",
    "library(SingleCellExperiment)\n",
    "library(dplyr)\n",
    "\n",
    "print(Sys.time())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b41e9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Same code per sample, please check sample ID and code in the \"SpaceX_sample_ID_sheet_file\" and manually change\n",
    "sample_ID <- \"ST37\" #corresponds to \"ST_ID column\"\n",
    "sample_name <- \"N2\" #corresponds to \"ST_name column\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42d6456",
   "metadata": {},
   "outputs": [],
   "source": [
    "#defining working directory\n",
    "indir = \"YOUR INPUT DIRECTORY HERE\"\n",
    "\n",
    "#define output directory\n",
    "outdir = \"YOUR OUTPUT DIRECTORY HERE\" \n",
    "\n",
    "setwd(indir)\n",
    "\n",
    "#directory with spatial coordinates,corresponds to \"Coordinates_per_sample\" folder in upload\n",
    "coordir = \"YOUR INPUT DIRECTORY HERE/Coordinates_per_sample/\"\n",
    "\n",
    "#directory with variable genes, corresponds to \"BayesSpace_highly_variable_genes\" folder in upload\n",
    "SVGdir = \"YOUR INPUT DIRECTORY HERE/BayesSpace_highly_variable_genes/\"\n",
    "\n",
    "#load the sample object\n",
    "print(paste0(\"Loading sample \", sample_ID))\n",
    "sample <- readRDS(paste0(indir, sample_ID, \"_group.rds\") )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a181d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(paste0(\"Loading coordinates\"))\n",
    "coordinates <- read.csv(paste0(coordir, sample_name, \"distance_to_lesioncore.csv\"), row.names = 1)\n",
    "coordinates$distance <- NULL\n",
    "BC_loc <- round(coordinates)\n",
    "BC_loc$barcodes <- substr(row.names(BC_loc), 1, 18) \n",
    "\n",
    "print(\"prepare gene expression matrix\")\n",
    "SVG.Selection <- read.csv(paste0(SVGdir, sample_ID, \"HVG_top500_fromBS.csv\"), row.names =1)\n",
    "expr_mat <- t(as.matrix(counts(sample))) \n",
    "# filter object for only the highly variable genes\n",
    "expr_mat <- expr_mat[, colnames(expr_mat) %in% SVG.Selection$x]\n",
    "\n",
    "print(\"removing empty spots that do not express the HVGs\")\n",
    "print(dim(expr_mat)[1])\n",
    "expr_mat <- expr_mat[rowSums(expr_mat) != 0,] #failsafe to remove spots with 0 genes expressed\n",
    "print(dim(expr_mat)[1])\n",
    "G <-dim(expr_mat)[2] # number of genes to model\n",
    "N <-dim(expr_mat)[1] # number of locations\n",
    "\n",
    "print(paste0(\"Number of genes is \", G, \"; Number of spots is \", N))\n",
    "\n",
    "matching_loc <- intersect(BC_loc$barcodes, rownames(expr_mat))\n",
    "print(paste0(\"Overlap barcodes in location and expression matrix is: \", length(matching_loc)))\n",
    "\n",
    "setdiff_loc <- setdiff(rownames(expr_mat), BC_loc$barcodes)\n",
    "print(paste0(\"Non-overlap barcodes in location and expression matrix is: \", length(setdiff_loc))) \n",
    "# this can be due to excluded spatial clusters\n",
    "\n",
    "BC_loc_test <- subset(BC_loc, BC_loc$barcodes %in% matching_loc)\n",
    "expr_mat_test <- subset(expr_mat, rownames(expr_mat) %in% matching_loc)\n",
    "BC_loc_test <- BC_loc_test[match(row.names(expr_mat_test), BC_loc_test$barcodes) , ]\n",
    "\n",
    "# load in cluster annotations: using 'identity' column which represent lesion core, rim, plwm instead of 'spatial.cluster'\n",
    "sample = sample[,rownames(expr_mat_test)]\n",
    "\n",
    "# check before adding info\n",
    "print('this should print TRUE, if FALSE investigate the filter step above')\n",
    "identical(colnames(sample), BC_loc_test$barcodes)\n",
    "\n",
    "# get information about the order in which networks in spatial clusters were created for later analyses\n",
    "BC_loc_test$cluster_ann <- sample$identity\n",
    "BC_loc_test$cluster_ann <- factor(BC_loc_test$cluster_ann)\n",
    "order_groups <- levels(BC_loc_test$cluster_ann)\n",
    "head(BC_loc_test)\n",
    "write.csv(order_groups, file = paste0(outdir, sample_ID, \"_orderofgroups.csv\"))\n",
    "\n",
    "#output files to be used in function for convenience and later post-processing\n",
    "saveRDS(expr_mat_test, paste0(outdir, sample_ID, \"_expr_mat_spacex.rds\"))\n",
    "saveRDS(BC_loc_test, paste0(outdir, sample_ID, \"_loc_spacex.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9f7bf80",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Post-process not performed in this part (ran separately) due to hardware limitations\n",
    "print(paste0(\"Starting SpaceX function, no postProcessing\"))\n",
    "Sys.time()\n",
    "BC_fit <- SpaceX(expr_mat_test,BC_loc_test[,1:2],BC_loc_test[,4], \n",
    "                 sPMM=FALSE, \n",
    "                 Post_process = FALSE, \n",
    "                 numCore=1) #\n",
    "Sys.time() %>% print()\n",
    "\n",
    "saveRDS(BC_fit, paste0(outdir, sample_ID, \"_spaceX_not_postProcessed.rds\") )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f1d6b5f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.2 (2021-11-01)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/bcn/p283607/anaconda3/envs/spacex/lib/libopenblasp-r0.3.21.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.3                 SingleCellExperiment_1.16.0\n",
       " [3] SummarizedExperiment_1.24.0 Biobase_2.54.0             \n",
       " [5] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        \n",
       " [7] IRanges_2.28.0              S4Vectors_0.32.4           \n",
       " [9] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       \n",
       "[11] matrixStats_1.1.0           SpaceX_0.1.0               \n",
       "[13] PQLseq_1.2.1               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] robust_0.7-1           Rcpp_1.0.11            MSFA_0.86             \n",
       " [4] mvtnorm_1.1-3          lattice_0.20-45        digest_0.6.33         \n",
       " [7] psych_2.3.3            foreach_1.5.2          utf8_1.2.4            \n",
       "[10] IRdisplay_1.1          R6_2.5.1               repr_1.1.6            \n",
       "[13] pcaPP_2.0-3            evaluate_0.23          pracma_2.4.2          \n",
       "[16] pillar_1.9.0           zlibbioc_1.40.0        rlang_1.1.2           \n",
       "[19] uuid_1.1-0             Matrix_1.6-2           statmod_1.5.0         \n",
       "[22] RCurl_1.98-1.13        DelayedArray_0.20.0    compiler_4.1.2        \n",
       "[25] pkgconfig_2.0.3        base64enc_0.1-3        mnormt_2.1.1          \n",
       "[28] htmltools_0.5.7        tidyselect_1.2.0       tibble_3.2.1          \n",
       "[31] GenomeInfoDbData_1.2.7 codetools_0.2-19       fansi_1.0.5           \n",
       "[34] rrcov_1.7-2            crayon_1.5.2           MASS_7.3-58.3         \n",
       "[37] bitops_1.0-7           grid_4.1.2             nlme_3.1-162          \n",
       "[40] jsonlite_1.8.7         lifecycle_1.0.4        magrittr_2.0.3        \n",
       "[43] cli_3.6.1              XVector_0.34.0         doParallel_1.0.17     \n",
       "[46] robustbase_0.95-1      generics_0.1.3         vctrs_0.6.4           \n",
       "[49] IRkernel_1.3.2         iterators_1.0.14       tools_4.1.2           \n",
       "[52] glue_1.6.2             DEoptimR_1.0-11        fit.models_0.64       \n",
       "[55] parallel_4.1.2         fastmap_1.1.1          pbdZMQ_0.3-9          \n",
       "[58] matlab_1.0.4          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.1.2 SpaceX",
   "language": "R",
   "name": "ir412_spacex"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
