{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8ec3b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# Author    : A. Alsema, J. Kotah\n",
    "# Date      : October 2023\n",
    "# Dataset   : Visium Spatial Transcriptomics for MS lesions, 14 WM samples\n",
    "# Purpose   : post-process files using SpaceX\n",
    "# Output    : RDS SpaceX analyzed object per sample that needs to be visualized and explored\n",
    "# Input     : RDS SpaceX analyzed object per sample that needs post-processing\n",
    "########################################################################\n",
    "\n",
    "# note: this part of the code was run based on the SpaceX post-processing source code, located here:\n",
    "# https://github.com/bayesrx/SpaceX/blob/main/R/SpaceX.R\n",
    "# please also note that we ran this on a high performance cluster due to intensive RAM requirements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7efacfa",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: PQLseq\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2024-02-28 14:35:56 CET\"\n"
     ]
    }
   ],
   "source": [
    "rm(list = ls())\n",
    "library(SpaceX)\n",
    "library(SingleCellExperiment)\n",
    "library(dplyr)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f23df23",
   "metadata": {},
   "source": [
    "### Example code, ran for all samples separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87a29bc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in SpaceX processed object from Script 2 for postprocessing\n",
    "fit_MSFA <- readRDS(\"20230707_ST33_test_spaceX_JK_not_postProcessed.rds\")\n",
    "\n",
    "#load in other SpaceX processed objects from Script 2\n",
    "Gene_expression_mat = readRDS(\"ST33_expr_mat_test.rds\")\n",
    "BC_loc_test = readRDS(\"ST33_loc_test.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "346c5c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "Spatial_locations = BC_loc_test[,1:2]\n",
    "Cluster_annotations = BC_loc_test[,4]\n",
    "\n",
    "nrun=10000 #default in SpaceX\n",
    "burn=5000 #default in SpaceX\n",
    "\n",
    "nrun = 10000 - 5000\n",
    "\n",
    "Spatial_loc = as.data.frame(cbind(Spatial_locations,Cluster_annotations))\n",
    "\n",
    "G <-dim(Gene_expression_mat)[2]\n",
    "L <- length(unique(Spatial_loc[,3]))\n",
    "Clusters <- unique(Spatial_loc[,3])\n",
    "\n",
    "Sys.time() %>% print()\n",
    "print(\"Start postprocessing\")\n",
    "\n",
    "SigmaPhi_post <- CorrPhi_post <- array(0, dim=c(G, G, nrun))\n",
    "Sigma_l_post <- Corr_l_post <- array(0, dim=c(G, G, nrun,L))\n",
    "\n",
    "  for(j in 1:nrun){\n",
    "    SigmaPhi_post[,,j] <- tcrossprod(fit_MSFA$Posterior_samples$Phi[,,j])\n",
    "    CorrPhi_post[,,j] <- cov2cor(SigmaPhi_post[,,j])\n",
    "    for (l in 1:L) {\n",
    "      Sigma_l_post[,,j,l] <- tcrossprod(fit_MSFA$Posterior_samples$Phi[,,j]) + tcrossprod(fit_MSFA$Posterior_samples$Lambda[[l]][,,j])\n",
    "      Corr_l_post[,,j,l] <- cov2cor(Sigma_l_post[,,j,l])\n",
    "    }\n",
    "  }\n",
    "\n",
    "print(\"Postprocessing done\")\n",
    "Sys.time() %>% print()\n",
    "\n",
    "Corr_l_est <- apply(Corr_l_post, c(1,2,4), mean)     ##Shared_network :: Shared co-expression matrix\n",
    "CorrPhi_est <- apply(CorrPhi_post, c(1,2), mean)  ## Cluster_network :: Cluster specific co-expression matrices\n",
    "\n",
    "AA <- list(Posterior_samples=fit_MSFA$Posterior_samples,\n",
    "           Shared_network=CorrPhi_est,\n",
    "           Cluster_network=Corr_l_est, \n",
    "           Genes = colnames(Gene_expression_mat)) #add gene names from matrix colnames as this is lost\n",
    "\n",
    "saveRDS(AA, \"008_ST33_SpaceX_Post-processed.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f1d6b5f0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.2 (2021-11-01)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/bcn/p283607/anaconda3/envs/spacex/lib/libopenblasp-r0.3.21.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] dplyr_1.1.3                 SingleCellExperiment_1.16.0\n",
       " [3] SummarizedExperiment_1.24.0 Biobase_2.54.0             \n",
       " [5] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        \n",
       " [7] IRanges_2.28.0              S4Vectors_0.32.4           \n",
       " [9] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       \n",
       "[11] matrixStats_1.1.0           SpaceX_0.1.0               \n",
       "[13] PQLseq_1.2.1               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] robust_0.7-1           Rcpp_1.0.11            MSFA_0.86             \n",
       " [4] mvtnorm_1.1-3          lattice_0.20-45        digest_0.6.33         \n",
       " [7] psych_2.3.3            foreach_1.5.2          utf8_1.2.4            \n",
       "[10] IRdisplay_1.1          R6_2.5.1               repr_1.1.6            \n",
       "[13] pcaPP_2.0-3            evaluate_0.23          pracma_2.4.2          \n",
       "[16] pillar_1.9.0           zlibbioc_1.40.0        rlang_1.1.2           \n",
       "[19] uuid_1.1-0             Matrix_1.6-2           statmod_1.5.0         \n",
       "[22] RCurl_1.98-1.13        DelayedArray_0.20.0    compiler_4.1.2        \n",
       "[25] pkgconfig_2.0.3        base64enc_0.1-3        mnormt_2.1.1          \n",
       "[28] htmltools_0.5.7        tidyselect_1.2.0       tibble_3.2.1          \n",
       "[31] GenomeInfoDbData_1.2.7 codetools_0.2-19       fansi_1.0.5           \n",
       "[34] rrcov_1.7-2            crayon_1.5.2           MASS_7.3-58.3         \n",
       "[37] bitops_1.0-7           grid_4.1.2             nlme_3.1-162          \n",
       "[40] jsonlite_1.8.7         lifecycle_1.0.4        magrittr_2.0.3        \n",
       "[43] cli_3.6.1              XVector_0.34.0         doParallel_1.0.17     \n",
       "[46] robustbase_0.95-1      generics_0.1.3         vctrs_0.6.4           \n",
       "[49] IRkernel_1.3.2         iterators_1.0.14       tools_4.1.2           \n",
       "[52] glue_1.6.2             DEoptimR_1.0-11        fit.models_0.64       \n",
       "[55] parallel_4.1.2         fastmap_1.1.1          pbdZMQ_0.3-9          \n",
       "[58] matlab_1.0.4          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.1.2 SpaceX",
   "language": "R",
   "name": "ir412_spacex"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
