{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c2a39050-2230-4b07-91f2-08ff520fc0b1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘monocle3’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    exprs, fData, fData<-, pData, pData<-\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "########################################################################\n",
    "# Author    : A. Alsema\n",
    "# Date      : April 2022\n",
    "# Dataset   : Visium Spatial Transcriptomics for MS lesions, 15 slices with WM \n",
    "# Purpose   : compute differentially expressed genes per selected subtrajectory\n",
    "\n",
    "# Required Inputs     : \n",
    "# - cds_objects: collection of CellDataSet (CDS) objects corresponding to various extracted subtrajectories. \n",
    "            # Each CDS object represents a subset of the overall trajectory focusing on specific paths,\n",
    "            # e.g. \"4-4.startnode_to_rims-microglia.rds\" contains a subset of spots belonging to subtrajectory 1, figure 6c-f.\n",
    "# - root_node: This variable represents the common starting point.\n",
    "\n",
    "# Outputs   : - new cds including subtrajectory with recalculated pseudotime,\n",
    "#             - csv file containing differential expression analysis per subtrajectory\n",
    "########################################################################\n",
    "\n",
    "rm(list = ls())\n",
    "library(monocle3)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "becc20f9-b9bf-4ad9-b632-4e992c2a71f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"subtraj_4-4.startnode_to_rims-microglia\"\n",
      "[1] \"subtraj data is saved, starting differential expression\"\n",
      "[1] \"subtraj_4-2.startnode_to_active_mixed\"\n",
      "[1] \"subtraj data is saved, starting differential expression\"\n",
      "[1] \"subtraj_4-1.startnode_to_infl\"\n",
      "[1] \"subtraj data is saved, starting differential expression\"\n"
     ]
    }
   ],
   "source": [
    "# this is the starting point of our trajectory, see figure 5 script 1\n",
    "root_node = readRDS(\"RData/monocle3/rootnode.rds\")\n",
    "\n",
    "# these cds paths corrrespond to an extracted subtrajectory, see figure 6 script 1\n",
    "cds_objects <- c(\"RData/monocle3/4-4.startnode_to_rims-microglia.rds\", \n",
    "                 \"RData/monocle3/4-2.startnode_to_active_mixed.rds\",\n",
    "                 \"RData/monocle3/4-1.startnode_to_infl.rds\")\n",
    "i = NULL\n",
    "\n",
    "for (i in 1:length(cds_objects)){\n",
    "    # loop over selected subtractories to calculate differential expression of genes using graph_test\n",
    "  cds_path = cds_objects[i]\n",
    "  cds_name = paste0(\"subtraj_\", substr(cds_path, 16, nchar(cds_path)-4))\n",
    "  print(cds_name)\n",
    "  cds_full <- readRDS(file = cds_path )\n",
    "  cds_full$sample_name <- NULL\n",
    "  \n",
    "    # recalculate pseudotime within subtrajectory.\n",
    "    # not sure if this is necessary\n",
    "  cds_full <- order_cells(cds_full, root_cells =  root_node) \n",
    "  \n",
    "    # save re-ordered subtrajectory\n",
    "  saveRDS(cds_full, file = paste0(\"RData/monocle3/\", cds_name, \"_graph.rds\"))\n",
    "  barcodes <- colnames(cds_full)\n",
    "  write.csv(barcodes, paste0(\"RData/monocle3/\", cds_name, \"_barcodes.csv\") )\n",
    "  \n",
    "    # compute differential expression\n",
    "  print(\"subtraj data is saved, starting differential expression\")\n",
    "  cds_pr_test_res <- graph_test(cds_full, neighbor_graph=\"knn\",\n",
    "                                verbose = FALSE, reduction_method = \"UMAP\", cores=1)\n",
    "                    ## bug in monocle3 when setting neighbor_graph = \"principal_graph\"\n",
    "                    ## here, opted for graph test with neighbor_graph = \"knn\" (is also the default)\n",
    "  write.csv(cds_pr_test_res, file = paste0(\"Routput/monocle3/trajectory_DE/1.DE_results_\", cds_name, \"_allgenes.csv\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7075262d-195e-4529-8c4e-1f12ad16981b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.2.0 (2022-04-22)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/bcn/p283607/anaconda3/envs/R4.2/lib/libopenblasp-r0.3.21.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] ggplot2_3.4.4               monocle3_1.3.1             \n",
       " [3] SingleCellExperiment_1.20.1 SummarizedExperiment_1.28.0\n",
       " [5] GenomicRanges_1.50.2        GenomeInfoDb_1.34.9        \n",
       " [7] IRanges_2.32.0              S4Vectors_0.36.2           \n",
       " [9] MatrixGenerics_1.10.0       matrixStats_1.1.0          \n",
       "[11] Biobase_2.58.0              BiocGenerics_0.44.0        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] ggrepel_0.9.4          Rcpp_1.0.11            lattice_0.20-45       \n",
       " [4] listenv_0.9.0          assertthat_0.2.1       digest_0.6.33         \n",
       " [7] utf8_1.2.4             parallelly_1.36.0      IRdisplay_1.1         \n",
       "[10] plyr_1.8.9             R6_2.5.1               repr_1.1.4            \n",
       "[13] evaluate_0.23          pillar_1.9.0           zlibbioc_1.44.0       \n",
       "[16] rlang_1.1.3            uuid_1.1-0             minqa_1.2.6           \n",
       "[19] nloptr_2.0.3           Matrix_1.6-3           labeling_0.4.3        \n",
       "[22] splines_4.2.0          lme4_1.1-34            igraph_1.5.1          \n",
       "[25] RCurl_1.98-1.12        munsell_0.5.0          proxy_0.4-27          \n",
       "[28] DelayedArray_0.24.0    compiler_4.2.0         pkgconfig_2.0.3       \n",
       "[31] base64enc_0.1-3        globals_0.16.2         htmltools_0.5.7       \n",
       "[34] tidyselect_1.2.0       gridExtra_2.3          tibble_3.2.1          \n",
       "[37] GenomeInfoDbData_1.2.9 codetools_0.2-18       viridisLite_0.4.2     \n",
       "[40] fansi_1.0.6            future_1.33.0          withr_3.0.0           \n",
       "[43] crayon_1.5.2           dplyr_1.1.4            MASS_7.3-58.1         \n",
       "[46] bitops_1.0-7           grid_4.2.0             nlme_3.1-160          \n",
       "[49] jsonlite_1.8.7         gtable_0.3.4           lifecycle_1.0.4       \n",
       "[52] magrittr_2.0.3         scales_1.3.0           cli_3.6.2             \n",
       "[55] farver_2.1.1           XVector_0.38.0         viridis_0.6.2         \n",
       "[58] generics_0.1.3         vctrs_0.6.5            boot_1.3-28           \n",
       "[61] IRkernel_1.3.1         tools_4.2.0            glue_1.7.0            \n",
       "[64] parallel_4.2.0         fastmap_1.1.1          terra_1.5-21          \n",
       "[67] colorspace_2.1-0       pbdZMQ_0.3-8          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R4.2",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
